### PR TITLE
[OSS-fuzz] Fix FuzzMongoRead

### PR DESCRIPTION
**Pull request by jakule**
After https://github.com/gravitational/teleport/pull/14609 was merged, OSS-Fuzz found a few more panics in Mongo driver/our implementation.
This PR contains fixes for all test cases that I was able to reproduce locally. Below's log shows a successful run using the code from that branch. Previously each run was causing panic in Go code or infinite memory allocation.

```
$ python3 infra/helper.py run_fuzzer teleport fuzz_mongo_read -- -rss_limit_mb=2560 -timeout=60 -runs=10000
INFO:root:Running: docker run --rm --privileged -i -e FUZZING_ENGINE=libfuzzer -e SANITIZER=address -e RUN_FUZZER_MODE=interactive -v /home/jnyckowski/projects/oss-fuzz/build/out/teleport:/out -t gcr.io/oss-fuzz-base/base-runner run_fuzzer fuzz_mongo_read -rss_limit_mb=2560 -timeout=60 -runs=10000.
Using seed corpus: fuzz_mongo_read_seed_corpus.zip
/out/fuzz_mongo_read -rss_limit_mb=2560 -timeout=25 -rss_limit_mb=2560 -timeout=60 -runs=10000 /tmp/fuzz_mongo_read_corpus < /dev/null
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 3508910774
INFO: 44561 Extra Counters
INFO:        1 files found in /tmp/fuzz_mongo_read_corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: seed corpus: files: 1 min: 16b max: 16b total: 16b rss: 36Mb
#2	INITED ft: 38 corp: 1/16b exec/s: 0 rss: 37Mb
#7	REDUCE ft: 38 corp: 1/13b lim: 16 exec/s: 0 rss: 37Mb L: 13/13 MS: 5 ChangeASCIIInt-ChangeBit-CopyPart-ChangeBinInt-EraseBytes-
#8	REDUCE ft: 38 corp: 1/10b lim: 16 exec/s: 0 rss: 37Mb L: 10/10 MS: 1 CrossOver-
#37	REDUCE ft: 38 corp: 1/8b lim: 16 exec/s: 0 rss: 37Mb L: 8/8 MS: 4 InsertByte-CrossOver-EraseBytes-EraseBytes-
#59	REDUCE ft: 38 corp: 1/7b lim: 16 exec/s: 0 rss: 37Mb L: 7/7 MS: 2 ShuffleBytes-EraseBytes-
#74	REDUCE ft: 38 corp: 1/4b lim: 16 exec/s: 0 rss: 37Mb L: 4/4 MS: 5 InsertByte-InsertByte-ChangeBinInt-EraseBytes-EraseBytes-
#86	REDUCE ft: 38 corp: 1/3b lim: 16 exec/s: 0 rss: 37Mb L: 3/3 MS: 2 ChangeASCIIInt-EraseBytes-
#92	REDUCE ft: 38 corp: 1/2b lim: 16 exec/s: 0 rss: 37Mb L: 2/2 MS: 1 EraseBytes-
#98	REDUCE ft: 39 corp: 2/3b lim: 16 exec/s: 0 rss: 37Mb L: 1/2 MS: 1 EraseBytes-
#111	NEW    ft: 40 corp: 3/7b lim: 16 exec/s: 0 rss: 37Mb L: 4/4 MS: 3 InsertRepeatedBytes-ChangeBinInt-ShuffleBytes-
#207	REDUCE ft: 40 corp: 3/5b lim: 16 exec/s: 0 rss: 37Mb L: 2/2 MS: 1 EraseBytes-
#332	REDUCE ft: 261 corp: 4/14b lim: 16 exec/s: 0 rss: 37Mb L: 9/9 MS: 5 InsertRepeatedBytes-ShuffleBytes-ChangeBinInt-CrossOver-ChangeBinInt-
#343	NEW    ft: 282 corp: 5/26b lim: 16 exec/s: 0 rss: 38Mb L: 12/12 MS: 1 InsertRepeatedBytes-
#1145	REDUCE ft: 393 corp: 6/46b lim: 20 exec/s: 0 rss: 39Mb L: 20/20 MS: 2 CMP-InsertRepeatedBytes- DE: "\020\000\000\000\000\000\000\000"-
#1174	NEW    ft: 394 corp: 7/65b lim: 20 exec/s: 0 rss: 39Mb L: 19/20 MS: 4 InsertRepeatedBytes-EraseBytes-PersAutoDict-CMP- DE: "\020\000\000\000\000\000\000\000"-"\000\000\000\000\000\000\000\022"-
#1192	REDUCE ft: 394 corp: 7/63b lim: 20 exec/s: 0 rss: 39Mb L: 18/19 MS: 3 CrossOver-ChangeBit-EraseBytes-
#1199	NEW    ft: 395 corp: 8/82b lim: 20 exec/s: 0 rss: 39Mb L: 19/19 MS: 2 ChangeASCIIInt-CMP- DE: "\377\377\377\025"-
#1235	NEW    ft: 396 corp: 9/101b lim: 20 exec/s: 0 rss: 39Mb L: 19/19 MS: 1 ChangeByte-
#1238	NEW    ft: 397 corp: 10/121b lim: 20 exec/s: 0 rss: 39Mb L: 20/20 MS: 3 ShuffleBytes-ShuffleBytes-InsertByte-
#1239	NEW    ft: 402 corp: 11/140b lim: 20 exec/s: 0 rss: 39Mb L: 19/20 MS: 1 ChangeByte-
#1248	REDUCE ft: 402 corp: 11/139b lim: 20 exec/s: 0 rss: 39Mb L: 18/20 MS: 4 ChangeASCIIInt-ChangeASCIIInt-PersAutoDict-EraseBytes- DE: "\377\377\377\025"-
#1285	NEW    ft: 403 corp: 12/159b lim: 20 exec/s: 0 rss: 39Mb L: 20/20 MS: 2 ChangeBit-CopyPart-
#1351	REDUCE ft: 404 corp: 13/177b lim: 20 exec/s: 0 rss: 39Mb L: 18/20 MS: 1 ChangeBit-
#1373	NEW    ft: 427 corp: 14/197b lim: 20 exec/s: 0 rss: 39Mb L: 20/20 MS: 2 ChangeByte-InsertByte-
#1394	REDUCE ft: 427 corp: 14/195b lim: 20 exec/s: 0 rss: 39Mb L: 18/20 MS: 1 EraseBytes-
#1406	NEW    ft: 429 corp: 15/215b lim: 20 exec/s: 0 rss: 40Mb L: 20/20 MS: 1 ChangeByte-
#1412	NEW    ft: 452 corp: 16/234b lim: 20 exec/s: 0 rss: 40Mb L: 19/20 MS: 1 PersAutoDict- DE: "\377\377\377\025"-
#1512	NEW    ft: 453 corp: 17/254b lim: 20 exec/s: 0 rss: 60Mb L: 20/20 MS: 2 ChangeByte-PersAutoDict- DE: "\020\000\000\000\000\000\000\000"-
#1619	NEW    ft: 455 corp: 18/272b lim: 20 exec/s: 0 rss: 60Mb L: 18/20 MS: 2 ShuffleBytes-EraseBytes-
#1682	NEW    ft: 460 corp: 19/291b lim: 20 exec/s: 0 rss: 60Mb L: 19/20 MS: 3 EraseBytes-ChangeBinInt-InsertRepeatedBytes-
#2641	REDUCE ft: 460 corp: 19/290b lim: 28 exec/s: 0 rss: 72Mb L: 18/20 MS: 1 CrossOver-
#2712	REDUCE ft: 460 corp: 19/289b lim: 28 exec/s: 0 rss: 72Mb L: 19/20 MS: 1 EraseBytes-
#2940	REDUCE ft: 460 corp: 19/288b lim: 28 exec/s: 0 rss: 72Mb L: 18/20 MS: 2 InsertByte-EraseBytes-
#3231	REDUCE ft: 460 corp: 19/286b lim: 28 exec/s: 0 rss: 72Mb L: 18/20 MS: 1 EraseBytes-
#3528	REDUCE ft: 460 corp: 19/285b lim: 28 exec/s: 0 rss: 72Mb L: 19/20 MS: 2 EraseBytes-ShuffleBytes-
#3724	REDUCE ft: 460 corp: 19/284b lim: 28 exec/s: 0 rss: 92Mb L: 18/20 MS: 1 EraseBytes-
#3971	REDUCE ft: 460 corp: 19/283b lim: 28 exec/s: 0 rss: 92Mb L: 18/20 MS: 2 EraseBytes-CopyPart-
#4035	REDUCE ft: 469 corp: 20/309b lim: 28 exec/s: 0 rss: 92Mb L: 26/26 MS: 4 CMP-CopyPart-ShuffleBytes-InsertRepeatedBytes- DE: "\005\000\000\000\000\000\000\000"-
#4225	REDUCE ft: 478 corp: 21/329b lim: 28 exec/s: 0 rss: 92Mb L: 20/26 MS: 5 ChangeByte-ChangeBit-InsertByte-CopyPart-ChangeBinInt-
#4258	REDUCE ft: 482 corp: 22/355b lim: 28 exec/s: 0 rss: 92Mb L: 26/26 MS: 3 ChangeBinInt-ShuffleBytes-CMP- DE: "\000\000\000\000\000\000\000\001"-
#4495	REDUCE ft: 482 corp: 22/351b lim: 28 exec/s: 0 rss: 92Mb L: 22/26 MS: 2 CrossOver-EraseBytes-
#4590	REDUCE ft: 482 corp: 22/350b lim: 28 exec/s: 0 rss: 92Mb L: 19/26 MS: 5 ChangeBinInt-PersAutoDict-ChangeASCIIInt-CrossOver-EraseBytes- DE: "\005\000\000\000\000\000\000\000"-
#4626	REDUCE ft: 562 corp: 23/372b lim: 28 exec/s: 0 rss: 92Mb L: 22/26 MS: 1 CMP- DE: "\021\000\000\000\000\000\000\000"-
#4712	NEW    ft: 563 corp: 24/394b lim: 28 exec/s: 0 rss: 92Mb L: 22/26 MS: 1 CrossOver-
#5418	NEW    ft: 565 corp: 25/422b lim: 32 exec/s: 0 rss: 92Mb L: 28/28 MS: 1 CopyPart-
#5619	REDUCE ft: 565 corp: 25/420b lim: 32 exec/s: 0 rss: 92Mb L: 26/26 MS: 1 EraseBytes-
#5662	REDUCE ft: 567 corp: 26/447b lim: 32 exec/s: 0 rss: 92Mb L: 27/27 MS: 3 InsertByte-ChangeBit-ShuffleBytes-
#6643	REDUCE ft: 567 corp: 26/446b lim: 37 exec/s: 0 rss: 92Mb L: 26/26 MS: 1 EraseBytes-
#8631	NEW    ft: 569 corp: 27/465b lim: 52 exec/s: 0 rss: 92Mb L: 19/26 MS: 3 ChangeBit-ShuffleBytes-ShuffleBytes-
#8926	REDUCE ft: 569 corp: 27/464b lim: 52 exec/s: 0 rss: 92Mb L: 25/26 MS: 5 CMP-InsertRepeatedBytes-ShuffleBytes-InsertByte-EraseBytes- DE: "\001\000\000\000\000\000\000\000"-
#9040	NEW    ft: 573 corp: 28/494b lim: 52 exec/s: 0 rss: 92Mb L: 30/30 MS: 4 ChangeBinInt-ChangeByte-ShuffleBytes-PersAutoDict- DE: "\000\000\000\000\000\000\000\001"-
#10000	DONE   ft: 573 corp: 28/494b lim: 57 exec/s: 0 rss: 92Mb
###### Recommended dictionary. ######
"\020\000\000\000\000\000\000\000" # Uses: 154
"\000\000\000\000\000\000\000\022" # Uses: 146
"\377\377\377\025" # Uses: 169
"\005\000\000\000\000\000\000\000" # Uses: 75
"\000\000\000\000\000\000\000\001" # Uses: 87
"\021\000\000\000\000\000\000\000" # Uses: 69
"\001\000\000\000\000\000\000\000" # Uses: 12
###### End of recommended dictionary. ######
Done 10000 runs in 0 second(s)
```

### CODE REVIEW

File: lib/srv/db/mongodb/protocol/opmsg.go [hunk 1]
```diff
@@ -274,6 +279,22 @@ func readOpMsg(header MessageHeader, payload []byte) (*MessageOpMsg, error) {
 	}, nil
 }
 
+// validateDocumentSize validates document length encoded in the message.
+func validateDocumentSize(src []byte) error {
+	if len(src) < 4 {
+		return trace.BadParameter("document is too short")
+	}
+	const headerLen = 4
```
Comment (smallinsky) [inline lib/srv/db/mongodb/protocol/opmsg.go:287]:
```suggestion
        const headerLen = 4
	if len(src) < headerLen {
		return trace.BadParameter("document is too short")
	}
	
```

File: lib/srv/db/mongodb/protocol/opreply.go [hunk 2]
```diff
@@ -138,6 +138,26 @@ func readOpReply(header MessageHeader, payload []byte) (*MessageOpReply, error)
...
+//
+// This function works in the same way as wiremessage.ReadReplyDocuments except, it can handle document of size 0.
+// When a document of size 0 is passed to wiremessage.ReadReplyDocuments, it will keep creating empty documents until
+// it uses all system memory/application crash.
+func ReadReplyDocuments(src []byte) (docs []bsoncore.Document, rem []byte, ok bool) {
+	rem = src
+	for {
+		var doc bsoncore.Document
+		doc, rem, ok = bsoncore.ReadDocument(rem)
+		if !ok || len(doc) == 0 {
```
Comment (smallinsky) [inline lib/srv/db/mongodb/protocol/opreply.go:151]: It looks reasonable but I think that we also need to  raise an issue in MongoDB driver lib to get this fix in the origin -  I will do that.

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/mongodb/protocol/opcompressed.go [hunk 3]
```diff
@@ -114,6 +114,10 @@ func readOpCompressed(header MessageHeader, payload []byte) (message *MessageOpC
 		CompressedMessage: compressedMessage,
 		bytes:             append(header.bytes[:], payload...),
 	}
+	if uncompressedSize <= 0 || len(compressedMessage) <= 0 {
```
Comment (zmb3) [inline lib/srv/db/mongodb/protocol/opcompressed.go:117]:
```suggestion
	if uncompressedSize <= 0 || len(compressedMessage) == 0 {
```

`len` will never return a negative number, right?

File: lib/srv/db/mongodb/protocol/opcompressed.go [hunk 4]
```diff
@@ -114,6 +114,10 @@ func readOpCompressed(header MessageHeader, payload []byte) (message *MessageOpC
...
+		return nil, trace.BadParameter("malformed OP_COMPRESSED: invalid message size %v", payload)
```
Comment (zmb3) [inline lib/srv/db/mongodb/protocol/opcompressed.go:118]: Should this error message print the size that is invalid, or the actual raw payload?

**Review by nklaassen — APPROVED** [review]

**Pull request merged**